### PR TITLE
[TKW] Implement define interface op and tk.max

### PR DESCRIPTION
**Pull request by raikonenfnu**
In this PR, we add 3 things:

    1. Define Interface Op
    2. Reduction max op
    3. elementwise maximum op

Define interface op is created to generate new subclass op that can
derive from base interface class. this is useful for custom TK ops that
has very similar properties such as "sum" and "max" S.T we can reuse
code as much as possible.

### CODE REVIEW

File: tests/kernel/wave/wave_e2e_test.py [hunk 1]
```diff
@@ -234,4 +234,4 @@ def test(
         run_config=config,
     ):
         test(a, b, c)
-        assert_allclose(ref, c, atol=0.1)
+        assert_allclose(ref, c, atol=0.15)
```
Comment (harsh-nod) [inline tests/kernel/wave/wave_e2e_test.py:237]: Why did the tolerance go up? This is doing a sum so it should have remained the same? Also can you add a max e2e test (for which I am guessing you should match exactly?) And not for this PR, but would be nice to compare against iree to see whether there are any differences (there should be none?)

**Review by harsh-nod — CHANGES_REQUESTED** [review]

File: tests/kernel/wave/wave_e2e_test.py (hunk 1 above)
Comment (raikonenfnu) [reply tests/kernel/wave/wave_e2e_test.py:237]: great Q, yeap, it is unrelated to this PR. But there was one instance the error gets up to 0.125, since test generate random values. it was for a really large number along the magnitude of 84.872 to 84.747. So I set 0.15 just to make sure it doesn't break ever. I can also set manual seed instead. :)

File: shark_turbine/kernel/wave/register_analysis.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 from .._support.tracing import CapturedTrace
 from ...support.logging import get_logger
-from ..ops.wave_ops import *
+from ..ops.wave_ops import get_custom, NewRegister, CustomOp, MMA, Reduction
+import torch.fx as fx
```
[the remark this answers is not in the document]
Comment (raikonenfnu) [reply shark_turbine/kernel/wave/register_analysis.py:4]: register_analysis uses "max", but since we import *, it imported tkw.max as "max" hahaha

File: shark_turbine/kernel/ops/wave_ops.py [hunk 3]
```diff
@@ -180,6 +192,53 @@ def new_function(*args: Any, **kwargs: dict[str, Any]):
...
+def define_interface_op(op_name: str) -> Callable[[T], T]:
+    """
+    Generate new subclass for op handling, deriving from the base interface class.
+    Generated subclass can be used for emitting the op from compiler/python side,
+    by calling the generated subclass name.
+
+    The generated subclass name would be pascal case of the TKW op name. For example:
+    "tkw.op_name" -> "OpName"
+    "tkw.exp2" -> "Exp2"
+    """
```
[the remark this answers is not in the document]
Comment (raikonenfnu) [reply shark_turbine/kernel/ops/wave_ops.py:204]: thanks! :)

File: shark_turbine/kernel/wave/register_analysis.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (harsh-nod) [reply shark_turbine/kernel/wave/register_analysis.py:4]: ah makes sense, thanks!

File: tests/kernel/wave/wave_e2e_test.py (hunk 1 above)
Comment (harsh-nod) [reply tests/kernel/wave/wave_e2e_test.py:237]: yeah let's set a manual seed so the tests results are reproducible.

Comment (raikonenfnu) [reply tests/kernel/wave/wave_e2e_test.py]: SG! let me push this up in a separate PR for better visibilityy and cleaner git commit history  :)

Comment (harsh-nod) [reply tests/kernel/wave/wave_e2e_test.py]: Sounds good. Have you validated that the max reduction works e2e?

Comment (raikonenfnu) [reply tests/kernel/wave/wave_e2e_test.py]: yeap! I have another repo where I collect and test TK-Kernels, LMK push it up here as well

Comment (raikonenfnu) [reply tests/kernel/wave/wave_e2e_test.py]: Added test, and this one doesn't even need the atol modified. 😁

Comment (harsh-nod) [reply tests/kernel/wave/wave_e2e_test.py]: nice! I think you can probably set to exact equality and it should still work because there is no loss in precision during max.

**Review by harsh-nod — APPROVED** [review]
thanks!

File: tests/kernel/wave/wave_e2e_test.py (hunk 1 above)
Comment (raikonenfnu) [reply tests/kernel/wave/wave_e2e_test.py]: ah shoot, missed this comment, I'll put it up in another PR fixing the reduce_sum test with manual seed. :)
